File: tads_lineares/c/filaCircular.c
#include <stdio.h>

#define MAXTAM 10

// Atributos
int array[MAXTAM];
int inicio, fim;

// Construtor
void start() { inicio = fim = 0; }

// Método de Inserção
void inserir(int num) {
  if ((fim + 1) % MAXTAM == inicio) {
    printf("Array Cheio");
  } else {
    array[fim] = num;
    fim = (fim + 1) % MAXTAM;
  }
}

// Método de Remoção
void remover() {
  if (fim == inicio) {
    printf("Array Vazio");
  } else {
    inicio = (inicio + 1) % MAXTAM;
  }
}

// Método de Exibição
void imprimir() {
  printf("[ ");
  for (int i = inicio; i < fim; i++) {
    printf("%d ", array[i]);
  }
  printf("]\n");
}

int main() {
  start();
  inserir(1);
  inserir(5);
  inserir(10);
  inserir(25);
  imprimir();
  remover();
  remover();
  remover();
  imprimir();
  return 0;
}

File: tads_lineares/java/FilaCircular.java
package tads_lineares.java;

public class FilaCircular {
    // Atributos
    private int[] array;
    private int inicio;
    private int fim;

    // Construtor
    public FilaCircular(int tamanho) {
        array = new int[tamanho + 1];
        inicio = fim = 0;
    }

    // Método de Inserção
    public void inserir(int num) throws Exception {
        if ((fim + 1) % array.length == inicio) {
            throw new Exception("Array Cheio");
        } else {
            array[fim] = num;
            fim = (fim + 1) % array.length;
        }
    }

    // Método de Remoção
    public void remover() throws Exception {
        if (fim == inicio) {
            throw new Exception("Array Vazio");
        } else {
            inicio = (inicio + 1) % array.length;
        }
    }

      // Método de Exibição
      public void imprimir() {
        System.out.print("[ ");
        for (int i = inicio; i < fim; i++) {
            System.out.print(array[i] + " ");
        }
        System.out.print("]");
    }
}
